(MODULES-5264) IIS: handle UNC paths

tkishel opened the pull request:
Prior to this commit, the IIS module:

included regular expressions that failed to a match a UNC paths
attempted to use File.exists? to test a UNC path
failed to set iis_virtual_directory physicalpath to a UNC path

Also, these issues interfered with development and testing:

ran iis_virtual_directory update() after destroy()
failed to set the iis_site :ensure attribute to :present
failed to set the iis_virtual_directory :application attribute

jpogran reviewed (changes requested):
We should merge #93 and #91 first, because this PR changes a lot of similar files.

This PR needs to be broken up into separate commits per provider

jpogran commented on lib/puppet/provider/iis_site/webadministration.rb:
```diff
@@ -171,4 +180,19 @@ def self.to_bool(value)
     raise ArgumentError.new("invalid value for Boolean: \"#{value}\"")
   end
+
+  def verify_physicalpath
```
We should move this to a helper method instead of having it in each provider

tkishel replied:
Could you recommend a file location?

jpogran commented on lib/puppet/provider/iis_site/webadministration.rb:
```diff
@@ -171,4 +180,19 @@ def self.to_bool(value)
+    end
+  end
+
+  def is_drive_path(path)
```
We should move this to a helper method instead of having it in each provider

jpogran commented on lib/puppet/provider/iis_virtual_directory/webadministration.rb:
```diff
@@ -106,11 +130,29 @@ def self.instances
+    return (path =~ /^.:(\/|\\)/)
+  end
+
+  def is_unc_path(path)
```
We should move this to a helper method instead of having it in each provider

jpogran commented on spec/unit/puppet/type/iis_application_pool_spec.rb:
```diff
@@ -113,7 +113,6 @@
   
   [
     :name,
-    :clr_config_file,
```
If this is removed, where is it added back to test?

tkishel replied:
Line 131.

Though I've identified a potential issue here: at least one 'path' attribute is documented as accepting variables:

https://docs.microsoft.com/en-us/iis/configuration/system.applicationhost/sites/sitedefaults/logfile/

`directory = %SystemDrive%\inetpub\logs\LogFiles`

I don't think this is true or common with physicalpath, but clr_config_file and source (iis_feature)?

jpogran commented on spec/unit/puppet/type/iis_site_spec.rb:
```diff
@@ -38,26 +38,6 @@
   end
 
-  # TODO: fix tests below
-  context "parameter :physicalpath" do
```
Can you explain why you removed the validation checks and tests for nil physical paths for websites? IIRC you need a physical path for a site

tkishel replied:
I'll restore that. Mislead by: 
https://technet.microsoft.com/en-us/library/hh867858(v=wps.630).aspx

tkishel replied:
```
-PhysicalPath<String>
Required? | false
```

jpogran commented on lib/puppet/provider/iis_site/webadministration.rb:
```diff
@@ -148,7 +152,12 @@ def self.instances
       end
 
-      site_hash[:ensure]               = site['state'].downcase
+      if site['state'] == ''
```
Why would state be nil here?

tkishel replied:
When `ensure => present`

```
  iis_site { "test_site":
    name            => "test_site",
    ensure          => 'present',
    physicalpath    => 'c:\\inetpub\\test_site',
    applicationpool => 'DefaultApplicationPool',
  }
```

```
PS C:\Users\Administrator> Get-WebSite
                                                                                                                                                                                                       
Name             ID   State      Physical Path                  Bindings
----             --   -----      -------------                  --------
Default Web Site 1    Started    %SystemDrive%\inetpub\wwwroot  http *:80: 
test_site        2               c:\inetpub\test_site           http *:80: 
```

tkishel commented:
I will need advice on reverting and breaking into separate commits per provider.

tkishel committed:
(MODULES-5264) IIS: handle UNC paths

Define path type and path helper methods.

verify_physicalpath
verify_optional_physicalpath
is_drive_path
is_unc_path

tkishel committed:
(MODULES-5264) IIS: handle UNC paths

Implement path type in iis_feature.
Fix desc typo.

tkishel committed:
(MODULES-5264) IIS: handle UNC paths

Implement path type in is_application_pool.

tkishel committed:
(MODULES-5264) IIS: handle UNC paths

Implement path type and path helper methods in iis_application.

tkishel committed:
(MODULES-5264) IIS: handle UNC paths

Implement path type and path helper methods in iis_site.
Correctly set the ensure attribute to present when Get-WebSite::State is empty.

tkishel committed:
(MODULES-5264) IIS: handle UNC paths

Implement path type and path helper methods in iis_virtual_directory.
Correctly set the application attribute.
Add parameter spec tests.

tkishel committed:
(MODULES-5264) IIS: handle UNC paths

Define require_path_for RSpec::Matcher.

tkishel commented:
I rebased and split my previous commit, created lib/puppet/provider/iis_common.rb for helper methods, and added spec tests. Please let me know if this is what you're expecting!

tkishel commented:
I could rebase this from master, or close and recreate, or back away slowly ... just let me know.

tkishel commented:
Closing in favor of a new PR (104) with a cleaner scope.
